### PR TITLE
MBS-9536 Add new ActivityScenarioRule

### PR DESCRIPTION
**Pull request by RuslanMingaliev**
Integration build:
https://tmct.avito.ru/viewLog.html?buildId=22863262&buildTypeId=AvitoAndroid_Integration

### CODE REVIEW

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/BitmapMatcherTest.kt [hunk 1]
```diff
@@ -53,5 +54,6 @@ class BitmapMatcherTest {
         assertThat(error).hasMessageThat().contains("expected=#FF0000FF")
     }
 
-    private fun getDrawable(@DrawableRes id: Int): Drawable = rule.activity.getDrawable(id)!!
+    private fun getDrawable(@DrawableRes id: Int): Drawable =
+        InstrumentationRegistry.getInstrumentation().targetContext.getDrawable(id)!!
```
Comment (sergioboi) [inline subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/BitmapMatcherTest.kt]: force unwrap to `requireNotNull() { "Can't find drawable by ${id to res name}" }

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ReadTextTest.kt [hunk 2]
```diff
@@ -22,8 +22,8 @@ class ReadTextTest {
     fun readNonBlankText_allowBlankIsFalse() = with(rule) {
         launchActivity(null)
 
-        rule.runOnUiThread {
-            activity.findViewById<EditText>(R.id.edit_text).setText("my_custom_text")
+        rule.onActivity {
+            findViewById<EditText>(R.id.edit_text).setText("my_custom_text")
```
Comment (sergioboi) [inline subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ReadTextTest.kt:26]: At what thread Is the lambda executed?

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/ActivityScenarioRule.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import com.avito.android.rule.base.BaseActivityScenarioRule
+
+class ActivityScenarioRule<T : Activity>(activityClass: Class<T>) : BaseActivityScenarioRule<T>(
+    activityClass = activityClass,
+    initialTouchMode = true,
+    launchActivity = false,
+) {
+
+    override fun afterActivityLaunched() {
+        Intents.init()
```
Comment (sergioboi) [inline subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/ActivityScenarioRule.kt:19]: What does that code do?

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import com.google.common.truth.Truth.assertThat
+import com.google.common.truth.Truth.assertWithMessage
+import org.junit.rules.TestRule
+import org.junit.runner.Description
+import org.junit.runners.model.Statement
+
+abstract class InHouseScenarioScreenRule<A : Activity>(activityClass: Class<A>) : TestRule {
+
+    protected val androidInstrumentation: Instrumentation = InstrumentationRegistry.getInstrumentation()
+    protected val arguments: Bundle = InstrumentationRegistry.getArguments()
```
Comment (sergioboi) [inline subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt]: Prefer to fetch arguments from instrumentation each time. They could be mutated

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        get() = activityRule.scenario.result
+
+    val scenario: ActivityScenario<A>
+        get() = activityRule.scenario
+
+    fun launchActivity(startIntent: Intent?): ActivityScenario<A> = activityRule.launchActivity(startIntent)
+
+    class ChecksLibrary<A : Activity>(private val scenarioFunc: () -> ActivityScenario<A>) {
+
+        fun isFinishing() {
```
Comment (sergioboi) [inline subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt]: let's rename to `isDestroyed` and `isNotDestroyed`

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/test/app/core/GodRuleChain.kt [hunk 6]
```diff
@@ -11,28 +12,22 @@ import org.junit.runners.model.Statement
...
     @Suppress("UNCHECKED_CAST")
     private val rules = chain.javaClass.getDeclaredField("rulesStartingWithInnerMost")
         .apply { isAccessible = true }.get(chain) as List<TestRule>
 
-    private val activityTestRule = rules.filterIsInstance<androidx.test.rule.ActivityTestRule<T>>().first()
+    private val activityTestRule = rules.filterIsInstance<BaseActivityScenarioRule<T>>().first()
 
-    val activity: T
-        get() = activityTestRule.activity
+    fun onActivity(func: T.() -> Unit): ActivityScenario<T> = activityTestRule.scenario.onActivity { func(it) }
```
Comment (sergioboi) [inline subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/test/app/core/GodRuleChain.kt:23]: Does this function have timeout or we will wait infinitely?

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/ActivityScenarioRule.kt (hunk 3 above)
Comment (eugene-krivobokov) [reply subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/ActivityScenarioRule.kt:19]:
```
/**
   * Initializes Intents and begins recording intents. Must be called prior to triggering any
   * actions that send out intents which need to be verified or stubbed. This is similar to
   * MockitoAnnotations.initMocks.
   */
   ```

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt (hunk 4 above)
Comment (eugene-krivobokov) [inline subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt]: Why do we need this property?
As I see, the rule doesn't use it itself and doesn't hide from children any information.

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt (hunk 5 above)
Comment (eugene-krivobokov) [reply subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt]: Add also please KDoc with a reference to Lifecycle state.

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ReadTextTest.kt (hunk 2 above)
Comment (RuslanMingaliev) [reply subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ReadTextTest.kt:26]: main

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt (hunk 4 above)
Comment (RuslanMingaliev) [reply subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt]: Just copied it from the currently used `InHouseScreenRule`.
Did not found usages in the Avito project too and removed the property.

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/test/app/core/GodRuleChain.kt (hunk 6 above)
Comment (RuslanMingaliev) [reply subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/test/app/core/GodRuleChain.kt:23]: Will wait for idle and perform the `func` on Activity

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/BitmapMatcherTest.kt (hunk 1 above)
Comment (RuslanMingaliev) [reply subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/BitmapMatcherTest.kt]: `requireNotNull` throws `IllegalArgumentException`. I think `Resources.NotFoundException` is more convenient in this case.

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt (hunk 4 above)
Comment (RuslanMingaliev) [reply subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt]: Removed the property at all

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        }
+
+        /**
+         * Asserts that activity is currently alive.
+         * For more information on lifecycle states, see [androidx.lifecycle.Lifecycle.State]
+         */
+        fun isNotDestroyed() {
+            waitForAssertion { assertThat(scenarioFunc().state).isNotEqualTo(Lifecycle.State.DESTROYED) }
+        }
+
```
Comment (eugene-krivobokov) [inline subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt:61]: Will clients need to check other states?
What about more generic way?

```kotlin
isInState(Lifecycle.State.RESUMED)
```

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/BitmapMatcherTest.kt [hunk 8]
```diff
@@ -53,5 +55,7 @@ class BitmapMatcherTest {
         assertThat(error).hasMessageThat().contains("expected=#FF0000FF")
     }
 
-    private fun getDrawable(@DrawableRes id: Int): Drawable = rule.activity.getDrawable(id)!!
+    private fun getDrawable(@DrawableRes id: Int): Drawable =
+        InstrumentationRegistry.getInstrumentation().targetContext.getDrawable(id)
+            ?: throw Resources.NotFoundException("Can't find drawable by id = $id")
```
Comment (sergioboi) [inline subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/BitmapMatcherTest.kt:60]: Better but optional convert id to an actual resource name

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/ActivityScenarioRule.kt (hunk 3 above)
Comment (sergioboi) [reply subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/ActivityScenarioRule.kt:19]: Why do we need that code like a default behavior?

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt (hunk 7 above)
Comment (RuslanMingaliev) [reply subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/InHouseScenarioScreenRule.kt:61]: Looks good to me. Replaced with 3 new methods

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/ActivityScenarioRule.kt (hunk 3 above)
Comment (RuslanMingaliev) [reply subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/ActivityScenarioRule.kt:19]: As I see, it just simplifies test writing with minimal performance penalties. Test authors don't copy this boilerplate code and have no chance to make mistakes, such as forgetting to call the `release` function.

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/BitmapMatcherTest.kt (hunk 8 above)
Comment (RuslanMingaliev) [reply subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/BitmapMatcherTest.kt:60]: I can't remember a simple way to convert id (just an integer value) to its resource name. It's possible by using reflection but seems like overengineering for this private function in the sample app.

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/ActivityScenarioRule.kt (hunk 3 above)
Comment (dsvoronin) [reply subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/ActivityScenarioRule.kt:19]: We don't. There is an issue(bug) about this,
but i believe it should be addressed in separate PR

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/base/BaseActivityScenarioRule.kt [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+import androidx.test.platform.app.InstrumentationRegistry
+import com.avito.android.rule.SimpleRule
+
+open class BaseActivityScenarioRule<A : Activity>(
+    private val activityClass: Class<A>,
+    private val initialTouchMode: Boolean,
+    private val launchActivity: Boolean,
+) : SimpleRule() {
+
+    private var _scenario: ActivityScenario<A>? = null
```
Comment (dsvoronin) [inline subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/base/BaseActivityScenarioRule.kt:16]: only cosmetics, but there should be a line separator between this var and `val scenario`

I don't remember we use underscore prefix anywhere in code

**Review by dsvoronin — APPROVED** [review]

File: subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/base/BaseActivityScenarioRule.kt (hunk 9 above)
Comment (RuslanMingaliev) [reply subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/base/BaseActivityScenarioRule.kt:16]: > only cosmetics, but there should be a line separator between this var and val scenario

done

> I don't remember we use underscore prefix anywhere in code

I used a [backing property](https://kotlinlang.org/docs/properties.html#backing-properties) syntax. I can rename if you're against it.

Comment (dsvoronin) [reply subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/base/BaseActivityScenarioRule.kt:16]: no, i'm not

**Comment by RuslanMingaliev** [comment]
@sboishtyan, going to merge this PR. Let me know if something is wrong

**Pull request merged**